File: android/app/src/main/java/la/aipnicmp/collector/collect/SpeedTest.kt
package la.aipnicmp.collector.collect

import android.util.Log
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import kotlin.math.roundToLong

/**
 * What the connection actually delivers, as opposed to what the bars claim.
 *
 * Signal strength is not speed. A phone can show four bars against a tower
 * whose backhaul is saturated, and a map built on RSRP alone would call that
 * village well served. Proposal 2.3 turns the points that carry both readings
 * into training labels for a model that estimates throughput everywhere else,
 * so these are the scarce, expensive samples the whole estimate rests on.
 *
 * Expensive literally: every run spends a collector's own mobile data. That
 * shapes everything here — the payload is small, the run is rare, and the
 * caller is expected to refuse far more often than it agrees.
 */
object SpeedTest {

    private const val TAG = "SpeedTest"

    /**
     * Bytes pulled per run. Small on purpose: 256 KB over a poor rural link is
     * already several seconds, and a larger sample would not change the verdict
     * for a map drawn in five colours.
     */
    const val PAYLOAD_BYTES = 262_144

    /** Round trips timed for latency. The best of three, not the mean — one
     *  scheduling hiccup on a busy handset should not become the finding. */
    private const val PING_ATTEMPTS = 3

    private const val CONNECT_TIMEOUT_MS = 10_000

    /**
     * Long enough for a genuinely slow link to finish, short enough that a
     * dying connection does not hold a foreground service open. A run that
     * times out reports nothing rather than a fabricated floor.
     */
    private const val READ_TIMEOUT_MS = 20_000

    data class Result(
        val downloadKbps: Double?,
        val latencyMs: Double?,
        /** Actually transferred, so the budget is charged what was spent. */
        val bytesUsed: Long,
    )

    fun run(baseUrl: String): Result {
        var spent = 0L
        val latency = measureLatency(baseUrl) { spent += it }
        val download = measureDownload(baseUrl) { spent += it }
        return Result(downloadKbps = download, latencyMs = latency, bytesUsed = spent)
    }

    private fun measureLatency(baseUrl: String, spend: (Long) -> Unit): Double? {
        var best: Double? = null
        repeat(PING_ATTEMPTS) {
            try {
                val connection = open("$baseUrl/api/v1/speedtest/ping")
                val started = System.nanoTime()
                connection.inputStream.use { it.readBytes() }
                val elapsed = (System.nanoTime() - started) / 1_000_000.0
                connection.disconnect()
                spend(64)
                if (best == null || elapsed < best!!) best = elapsed
            } catch (problem: Exception) {
                Log.w(TAG, "ping failed: ${problem.message}")
            }
        }
        return best
    }

    private fun measureDownload(baseUrl: String, spend: (Long) -> Unit): Double? {
        return try {
            val connection = open("$baseUrl/api/v1/speedtest/payload?bytes=$PAYLOAD_BYTES")
            // Timed from the first byte, not from connect: DNS, TCP and TLS are
            // real costs but they are latency, already measured above, and
            // counting them again here would depress the throughput figure on
            // exactly the weak links this project is looking for.
            connection.inputStream.use { stream ->
                val first = stream.read()
                if (first < 0) return null
                val started = System.nanoTime()
                val read = drain(stream) + 1
                val seconds = (System.nanoTime() - started) / 1_000_000_000.0
                connection.disconnect()
                spend(read.toLong())
                if (seconds <= 0.0 || read <= 0) null else (read * 8.0 / 1000.0) / seconds
            }
        } catch (problem: Exception) {
            Log.w(TAG, "download failed: ${problem.message}")
            null
        }
    }

    private fun drain(stream: InputStream): Int {
        val buffer = ByteArray(16 * 1024)
        var total = 0
        while (true) {
            val n = stream.read(buffer)
            if (n < 0) break
            total += n
        }
        return total
    }

    private fun open(url: String): HttpURLConnection =
        (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = CONNECT_TIMEOUT_MS
            readTimeout = READ_TIMEOUT_MS
            useCaches = false
            // Compression would measure how fast the phone inflates zeros. The
            // server refuses it too; asking as well means a proxy in between
            // cannot quietly turn a rural link into a fast one.
            setRequestProperty("Accept-Encoding", "identity")
            setRequestProperty("Cache-Control", "no-store")
        }

    /** Rounded so a stored figure never implies more precision than a single
     *  256 KB sample over a moving handset can support. */
    fun round(value: Double?): Double? = value?.let { (it * 10).roundToLong() / 10.0 }
}

File: backend/app/api/v1/speedtest.py
"""Endpoints a collector uses to measure the connection it is standing in.

Signal strength is not speed. A phone can show four bars against a tower whose
backhaul is saturated, and a coverage map built only on RSRP would call that
place well served. Proposal 2.2 asks for both, and section 2.3 depends on it:
the points carrying a real throughput reading are the training labels for the
model that estimates speed everywhere else from radio alone. Without any of
them that relationship can never be learned.

Served from this platform rather than a public speed-test host. It measures the
path the platform actually cares about — collector to this API — keeps the
bytes on infrastructure the ministry controls, and adds no third party to a
request made from a government project's handset.
"""

from __future__ import annotations

import os

from fastapi import APIRouter, Query, Response

router = APIRouter(prefix="/speedtest", tags=["speedtest"])

# The largest payload a collector may ask for.
#
# Sized for the connection this is aimed at, not for a fast one: 1 MB over a
# poor rural link is already several seconds of a collector's data allowance,
# and a bigger sample would not change the answer. Anything larger is refused
# rather than clamped, so a client asking for it learns that it was wrong.
MAX_PAYLOAD_BYTES = 1_048_576
DEFAULT_PAYLOAD_BYTES = 262_144


@router.get("/ping")
async def ping() -> Response:
    """The smallest possible reply, for timing a round trip.

    Deliberately not /health: that touches the database, so its timing would
    include work no collector is trying to measure.
    """
    return Response(
        content=b"1",
        media_type="text/plain",
        headers={"Cache-Control": "no-store", "Content-Length": "1"},
    )


@router.get("/payload")
async def payload(
    bytes_: int = Query(
        default=DEFAULT_PAYLOAD_BYTES,
        alias="bytes",
        ge=1024,
        le=MAX_PAYLOAD_BYTES,
        description="How many bytes to send back.",
    ),
) -> Response:
    """A block of incompressible bytes, for timing a download.

    Random rather than a repeated pattern, and served with compression refused:
    a payload that gzips to nothing would measure how fast the phone can
    inflate zeros, and report a rural village as having excellent throughput.

    Uncacheable for the same reason — a cached second run would time the
    phone's own storage.
    """
    return Response(
        content=os.urandom(bytes_),
        media_type="application/octet-stream",
        headers={
            "Cache-Control": "no-store, no-cache, must-revalidate",
            "Content-Encoding": "identity",
            "Content-Length": str(bytes_),
        },
    )
